Introduce a lookup endpoint to retrieve the Feed ID from a (profile) URL

pfefferle opened the pull request:
A new endpoint `/reader/feeds/lookup` allows resolving a Feed ID from a feed or profile URL.

The previous approach relied on `/reader/feeds/:feed_id`, later expanded to `/reader/feeds/*` with middleware mapping the wildcard to `feed_id`. While functional for many feeds, this design failed when handling URLs with slashes or special characters. For example, encoded Mastodon profile URLs such as `mastodon.social/@pfefferle` were misinterpreted, causing the router to treat part of the path as the `feed_id` rather than a complete URL.

The dedicated lookup endpoint correctly supports full (encoded) feed and profile URLs, including Fediverse accounts. This improves compatibility with ActivityPub integration, ensuring that following a Fediverse user automatically resolves the correct RSS feed and directs WordPress users to the appropriate feed afterwards.

## Proposed Changes

* Add dedicated /reader/feeds/lookup endpoint to resolve Feed ID by URL
* Introduce lookup endpoint for mapping feed/profile URLs to Feed IDs
* Provide dedicated lookup endpoint to fetch Feed ID from full (encoded) URLs

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* We need to look up Feeds by the Profile URLs of ActivityPub/Fediverse Users

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to https://wordpress.com/reader/subscriptions and subscribe to `https://mastodon.social/@pfefferle`
* The following URL will redirect you to the Feed in the Reader then: http://calypso.localhost:3000/reader/feeds/lookup/https://mastodon.social/@pfefferle

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

pfefferle committed:
Support wildcard feed routes in reader

Updated the feed route from '/reader/feeds/:feed_id' to '/reader/feeds/*' and added middleware to map the wildcard to the feed_id parameter. This change improves compatibility with feed URLs that may contain slashes or special characters.

Addison-Stavlo reviewed (commented):
> The current change would support full (encoded) URLs, but I fear to break something else.

I think this breaks the `/reader/feeds/{feedId}/posts/{postId}` route. 🤔

jeherve reviewed (commented):
> Calypso can look up feeds by URL

Do you know if that's something that is actually used anywhere in Calypso? As far as I can tell, we always load feeds using their IDs.
I'm no expert of the Reader though, so I may be missing a use-case.

> we want to subscribe to that user’s RSS feed. To direct the WordPress user to the correct feed afterwards, we need to resolve it by URL.

Would it be an option for the subscribe action to return the feed ID, so you could use that in the redirect URL?

pfefferle committed:
Add feed discovery by URL route to reader

Introduces a new route '/reader/feeds/lookup/*' that allows feed discovery by URL. Adds the corresponding controller logic, updates the route configuration, and registers the new section for proper module loading and performance tracking.

pfefferle committed:
Rename feedDiscoveryByUrl to feedLookup

Renamed the feedDiscoveryByUrl function to feedLookup in both the controller and route registration to improve naming consistency and clarity.

Addison-Stavlo commented on client/sections.js:
```diff
@@ -409,6 +409,14 @@ const sections = [
+	{
+		name: 'reader',
+		paths: [ '/reader/feeds/lookup/.*' ],
+		module: 'calypso/reader/full-post',
```
Im not entirely sure what makes sense or is needed here, considering we are only/always redirecting to either the feed page or following. But full-post is also neither of those.

Addison-Stavlo replied:
Oh probably just calypso/reader

pfefferle replied:
oops! copy and paste issue 🫣

pfefferle committed:
Update module path for reader section

Changed the module for the 'reader' section from 'calypso/reader/full-post' to 'calypso/reader' to ensure the correct module is loaded for the specified paths.

pfefferle committed:
Merge branch 'trunk' into fix/feed-url-lookup

Addison-Stavlo commented:
Sounds good! No problem and good luck with the other approach.

pfefferle committed:
Merge branch 'trunk' into fix/feed-url-lookup

pfefferle committed:
Remove redundant check for context.params[0] in feedLookup

Eliminated an unnecessary conditional that checked for the existence of context.params[0] before proceeding in the feedLookup function. The function now assumes the parameter is present, streamlining the code.

pfefferle committed:
Redirect to /reader if feed param is missing

Adds a check in feedLookup to redirect to /reader when the required feed parameter is not present in the context, preventing errors from undefined parameters.

pfefferle commented:
@obenland any ideas on that?

pfefferle committed:
Merge branch 'trunk' into fix/feed-url-lookup

obenland commented on client/reader/index.ts:
```diff
@@ -103,6 +104,16 @@ export default async function (): Promise< void > {
+			feedLookup,
+			feedListing,
+			makeLayout,
+			clientRender
```
Do we need anything other than `feedLookup` in here? It could help with performance to remove any unneeded items. Unfortunately, even when everything is removed, it still does a sites request, which in my case takes forever :)

pfefferle replied:
I think we can remove them... I had an error message when started working on that, that needed these, but now with only redirects, I think it is save to remove them.

jeherve commented on client/reader/controller.js:
```diff
@@ -121,6 +121,28 @@ export function feedDiscovery( context, next ) {
+	const url = context.params[ 0 ];
+	context.queryClient
+		.fetchQuery( {
+			queryKey: [ 'feed-discovery', url ],
```
You could probably update that key to match the new function.

jeherve commented on client/reader/controller.js:
```diff
@@ -121,6 +121,28 @@ export function feedDiscovery( context, next ) {
 	}
 }
 
+export function feedLookup( context ) {
```
Any chance you could add a docblock, to explain a bit why we're adding this new method and what it's used for?

jeherve reviewed (commented):
> Subscribe to https://mastodon.social/@pfefferle

Where would I do that? In specific pages of the Reader?

pfefferle commented:
> Where would I do that? In specific pages of the Reader?

The classic subscription process on https://wordpress.com/reader/subscriptions

pfefferle committed:
Add JSDoc and update queryKey in feedLookup

Added a JSDoc comment to the feedLookup function for better documentation. Also updated the queryKey from 'feed-discovery' to 'feed-lookup' to better reflect the function's purpose.

pfefferle committed:
Merge branch 'trunk' into fix/feed-url-lookup

pfefferle committed:
Simplify feed lookup route handlers

The '/reader/feeds/lookup/*' route now only uses 'redirectLoggedOutToSignup' and 'feedLookup' handlers, removing sidebar, feedListing, makeLayout, and clientRender. This streamlines the route's processing and may improve performance or maintainability.

pfefferle committed:
Merge branch 'fix/feed-url-lookup' of https://github.com/Automattic/wp-calypso into fix/feed-url-lookup

jeherve reviewed (approved):
This looks good and tests well for me now.
